apps/src/applab/ChartApi.js

```diff
@@ -0,0 +1,217 @@
+
+/** @enum {string} */
+ChartApi.ChartType = {
+  BAR: 'bar',
+  PIE: 'pie',
+  LINE: 'line',
+  SCATTER: 'scatter'
+};
+
+/** @type {Object.<string, GoogleChart>} */
+ChartApi.TypeNameToType = {
+  'bar': GoogleChart.MaterialBarChart,
+  'pie': GoogleChart.PieChart,
+  'line': GoogleChart.MaterialLineChart,
+  'scatter': GoogleChart.MaterialScatterChart
+};
+
+/**
+ * Get an array of all the chart type strings.
+ * @returns {string[]}
+ */
+ChartApi.getChartTypeNames = function () {
+  return Object.getOwnPropertyNames(ChartApi.TypeNameToType);
+};
+
+/**
+ * @param {ChartType} chartType
+ * @returns {boolean} TRUE if the given type is in the known list of chart types.
+ */
+ChartApi.supportsType = function (chartType) {
+  return ChartApi.getChartTypeNames().indexOf(chartType.toLowerCase()) !== -1;
+};
+
+/**
+ * @return {string[]} a quoted, sorted list of chart types for use in the
+ *         Droplet parameter dropdown.
+ */
+ChartApi.getChartTypeDropdown = function () {
+  return ChartApi.getChartTypeNames().map(quote).sort();
+};
+
+function quote(str) {
+  return '"' + str + '"';
+}
+
+/**
+ * Render a chart into an Applab chart element.
+ * @param {string} chartId - ID of the destination chart element.
+ * @param {ChartType} chartType - Desired chart type.
+ * @param {Object[]} chartData - Data to populate the chart.
+ * @param {Object} options - passed through to the API.
+ * @returns {Promise} which resolves when the chart has been rendered, or
+ *          rejects if there are any problems along the way.
+ */
+ChartApi.prototype.drawChart = function (chartId, chartType, chartData, options) {
+  try {
+    var chart = this.createChart_(chartId, chartType);
+    chart.loadData(chartData).then(function () {
+      return chart.drawChart(options);
+    }).then(function () {
+      this.mergeWarnings_(chart.warnings);
+    }.bind(this));
+  } catch (e) {
+    return Promise.reject(e);
+  }
+};
+
+/**
+ * Render a chart into an Applab chart element using data from an AppStorage
+ * API table.
+ * @param {string} chartId - ID of the destination chart element.
+ * @param {ChartType} chartType - Desired chart type.
+ * @param {string} tableName - AppStorage API table name to source data from
+ *                 for the chart.
+ * @param {string[]} columns - Columns to use from the table data for the chart,
+ *        in order (required order dependent on chart type).
+ * @param {Object} options - passed through to the API.
+ * @returns {Promise} resolves when the chart has been rendered, or rejects if
+ *          there are any problems along the way.
+ */
+ChartApi.prototype.drawChartFromRecords = function (chartId, chartType,
+    tableName, columns, options) {
+  try {
+    var chart = this.createChart_(chartId, chartType);
+    return Promise.all([
+      chart.loadDependencies(),
+      this.fetchTableData_(tableName)
+    ]).then(function (results) {
+      return chart.loadData(results[1], columns);
```

Comment:
That's better but still not great, because then students are prevented from explicitly including `id` as a column. I would recommend extracting the call to inferColumnsFromData out of loadData and into `ChartApi.prototype.drawChartFromRecords` here.